src/operators.rs

```diff
@@ -631,399 +743,320 @@ where
+                                    on_eof_histogram,
+                                    labels,
+                                    unwrap_any!(state.on_eof(py, &key))
+                                );
+
+                                // Update metrics.
+                                item_out_count.add(output.len() as u64, &labels);
+
+                                // Send items downstream.
+                                for value in output {
+                                    kv_downstream_session.give((key.clone(), TdPyAny::from(value)));
+                                }
+
+                                // Keep track of the fact we need to discard this.
+                                // We'll do this later on for all the keys at once.
+                                if let IsComplete::Discard = is_complete {
+                                    discarded_keys.push(key.clone());
                                 }
+
+                                awoken_keys_buffer.insert(key.clone());
                             }
+                        });
+
+                        for key in discarded_keys {
+                            state.remove(&key);
+                            sched_cache.remove(&key);
                         }
+                    }
 
-                        let load_frontier = &input_frontiers[1];
-                        if load_frontier.is_eof() {
-                            // Since we might emit downstream without any incoming
-                            // items, like on window timeout, ensure we FFWD to the
-                            // resume epoch.
-                            init_caps.downgrade_all(&resume_epoch.0);
+                    // Then go through all awoken keys and update the next scheduled
+                    // notification times.
+                    Python::with_gil(|py| {
+                        for key in awoken_keys_buffer.iter() {
+                            if let Some(sched) = with_timer!(
+                                notify_at_histogram,
+                                labels,
+                                unwrap_any!(state.notify_at(py, key))
+                            ) {
+                                sched_cache.insert(key.clone(), sched);
+                            }
                         }
+                    });
 
-                        // Schedule operator activation at the soonest
-                        // requested notify at for any key.
-                        if let Some(next_notify_at) =
-                            sched_cache.values().map(|notify_at| *notify_at - now).min()
-                        {
-                            activator.activate_after(
-                                next_notify_at.to_std().unwrap_or(std::time::Duration::ZERO),
-                            );
+                    // Snapshot and output state changes.
+                    if !awoken_keys_buffer.is_empty() {
```

Comment:
I think it ends up behaving the same way [1], but yeah the condition checks is unnecessarily convoluted and the snapshotting code duplicated, fixing this.

[1] `awoken_keys_buffer` is cleared if it had any elements and:
- At each activation if `immediate_snapshot` is on
- On epoch closing otherwise
But it only works because `immediate_snapshot` is set to `false` if recovery is `None`